server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java

```diff
@@ -125,21 +126,37 @@ public static long sizeOfRelocatingShards(
 
         // Where reserved space is unavailable (e.g. stats are out-of-sync) compute a conservative estimate for initialising shards
         for (ShardRouting routing : node.initializing()) {
-            if (routing.relocatingNodeId() == null) {
-                // in practice the only initializing-but-not-relocating shards with a nonzero expected shard size will be ones created
+            if (routing.relocatingNodeId() == null && metadata.index(routing.index()).isSearchableSnapshot() == false) {
+                // in practice the only initializing-but-not-relocating non-searchable-snapshot shards with a nonzero expected shard size
+                // will be ones created
                 // by a resize (shrink/split/clone) operation which we expect to happen using hard links, so they shouldn't be taking
                 // any additional space and can be ignored here
                 continue;
             }
             if (reservedSpace.containsShardId(routing.shardId())) {
                 continue;
             }
-
             final String actualPath = clusterInfo.getDataPath(routing);
             // if we don't yet know the actual path of the incoming shard then conservatively assume it's going to the path with the least
             // free space
             if (actualPath == null || actualPath.equals(dataPath)) {
-                totalSize += getExpectedShardSize(routing, 0L, clusterInfo, null, metadata, routingTable);
+                totalSize += getExpectedShardSize(
+                    routing,
+                    Math.max(routing.getExpectedShardSize(), 0L),
+                    clusterInfo,
+                    null,
+                    metadata,
+                    routingTable
+                );
+            }
+        }
+
+        // Count the STARTED searchable snapshot shards which are unaccounted in the cluster
+        for (ShardRouting shard : node.started()) {
```

Comment:
I am a bit afraid of the time complexity here. With 1000 started shards here, this turns a bit expensive and runs many times on every reroute.

I'd propose to cache this in `RoutingAllocation` such that we cache the amount of total size to add by `RoutingNode` but also keep a "modification counter" from the `RoutingNode` in the cache to detect that it changed. We could also more explicitly clear the cache upon changes but that seems like a more advanced solution.

Reply:
Hey @henningandersen! I created a static cache in `RoutingAllocation` based on `RoutingNode`. If `RoutingNode` changes, the cache will calculate a new value because `equals&hasCode` will be changed too. I also added a TTL for the cache, so we update the cache based on fresh `ClusterServiceInfo`